components/server/src/user/env-var-service.ts

```diff
@@ -265,7 +267,16 @@ export class EnvVarService {
             merge(UserEnvVar.filter(userEnvVars, wsContext.repository.owner, wsContext.repository.name));
         }
 
-        // 2. then from the project
+        // 2. then from the .gitpod.yml
```

Comment:
Sorry to open uo this discussion again - but I really don't recall.

Why don't we have it at the bottom, before user env vars? :thinking:

Reply:
Sorry for the lack of explanation, on this I've actually circled back and ended up doing differently than we talked about sync.

The reason is that I took a look at different scenarios you could hit as a user, and concluded prioritizing these global environment variables from them `.gitpod.yml` makes them more easily recoverable. Take the following:

The repository defines a variable, which is somehow essential to running the project. If a user defines a variable of the same name in their user settings, they could easily override this variable, essential for the project. With that said, overriding for sure could be desirable. But, if it's not, we do not offer any way (that I'm aware of) allowing the user to _exclude_ a single repository from picking up a user env. Once you scope to */*, it stays like that. 

If the order is flipped, you are able to override the repo environment variable either by actually changing it in the source code, or importing the repository and using Repository-scopes variables for the offending variable.

To sum it up, in the end, it made more sense to me that these more specific variables override the less specific user-scoped ones[^1]. The preference for the UI we introduce here (Repo config variables override the `.gitpod.yml` values) should therefore only apply for this case because they are on a similar level (although it still definitely can be argued that an imported repository is more specific than the underlying git repo, because there can be multiple stemming from the same one).

[^1]: there's a small thing that I did not consider with this, and that being user-scoped variables that are a perfect match (e.g., `gitpod-io/gitpod` as opposed to `gitpod-io/*`). Those are the same specificity as the ones defined in `.gitpod.yml` and arguable should override those values, because in this context, there is some inherently added specificity because it's on the user level. We could also consider this, but it would make the changes a bit more complex.

Reply:
We met today and agreed to switch the order to make the layers here easier to understand (we prioritize meta always).

Reply:
Changed in [`794d0aa` (#20339)](https://github.com/gitpod-io/gitpod/pull/20339/commits/794d0aaba02650658c5d5a6e1c08fd59028d2f87)